src/frequenz/sdk/timeseries/periodic_feature_extractor.py

```diff
@@ -0,0 +1,443 @@
+        given the distance between two window neighbors.
+
+        Args:
+            window: The buffer that is used for the average calculation.
+            window_size: The size of the window in samples.
+
+        Returns:
+            The number of windows that are fully contained in the MovingWindow.
+        """
+        num_windows = math.floor(len(window) / self._distance)
+        if len(window) - num_windows * self._distance >= window_size:
+            num_windows += 1
+
+        return num_windows
+
+    def _reshape_np_array(
+        self, array: NDArray[np.float_], window_size: int
+    ) -> NDArray[np.float_]:
+        """
+        Reshape a numpy array to a 2D array where each row represents a window.
+
+        There are three cases to consider
+
+        1. the array size is a multiple of window_size + distance,
+           i.e. num_windows is integer and we can simply reshape
+        2. the array size is NOT a multiple of window_size + distance and
+           it has length n * (window_size + distance) + m, where m < window_size
+        3. the array size is NOT a multiple of window_size + distance and
+           it has length n * (window_size + distance) + m, where m >= window_size
+
+        Args:
+            array: The numpy array to reshape.
+            window_size: The size of the window in samples.
+
+        Returns:
+            The reshaped 2D array.
+
+        Raises:
+            ValueError: If the array is smaller or equal to the given distance.
+        """
+        # Not using the num_windows function here because we want to
+        # differentiate between the three cases.
+        if len(array) < self._distance:
+            raise ValueError(
+                f"The array (length:{len(array)}) is too small to be reshaped."
+            )
+
+        num_windows = math.floor(len(array) / self._distance)
+
+        # Case 1:
+        if len(array) - num_windows * self._distance == 0:
+            log.debug("Case 1:")
+            resized_array = array
+        # Case 2
+        elif len(array) - num_windows * self._distance < window_size:
+            log.debug("Case 2:")
+            resized_array = array[: num_windows * self._distance]
+        # Case 3
+        else:
+            log.debug("Case 3:")
+            num_windows += 1
+            resized_array = np.resize(array, num_windows * self._distance)
+
+        return resized_array.reshape(num_windows, self._distance)
+
+    def _calculate_avg_window_py(
+        self,
+        window: NDArray[np.float_],
+        window_size: int,
+        weights: List[float] | None = None,
+    ) -> NDArray[np.float_]:
+        """
+        Plain python version of the average calculator.
+
+        This method avoids copying in any case but is 15 to 600
+        slower then the numpy version.
+
+        This method is only used in benchmarks and unit tests.
+
+        ONLY USE THIS IF YOU NEED TO STRICTLY AVOID COPIES!
```

Comment:
You can use special blocks to highlight this, for example:

```suggestion
        Danger:
            This method avoids copying in any case but is 15 to 600
            slower then the numpy version.

            This method is only used in benchmarks and unit tests.

            **ONLY USE THIS IF YOU NEED TO STRICTLY AVOID COPIES!**
```

(you can use `Warning:` too)

Here is the output:

![image](https://user-images.githubusercontent.com/108664423/235879231-ff6d417e-ca01-4ab2-8369-f7e0cdd6e14e.png)

![image](https://user-images.githubusercontent.com/108664423/235879453-6467b921-14bb-436b-94a6-f9506e54553a.png)

The only downside is right now this is rendered collapsed, not expanded, but there is an [issue](https://github.com/mkdocstrings/mkdocstrings/issues/468) in `mkdocstrings` to fix this (and I think it should be a pretty easy so I guess it should come eventually).

Reply:
I just noticed this is private, so it won't be actually rendered, but in any case I think is good to always write docs as if they were public, in case sometime we decide to make it public, as it happened in this PR :D

Reply:
The `Note` and `Note further` in the class docstring can probably use this pattern as well?

Reply:
I missed those, but yeah, there is a `Note:` *admonition* too. You can see the full list here: https://squidfunk.github.io/mkdocs-material/reference/admonitions/#supported-types

Reply:
And now admonitions are expanded by default! :tada:

https://github.com/mkdocstrings/mkdocstrings/issues/468#issuecomment-1537489843